### PR TITLE
Adds the ability to save/load circuits for admins. Adds the ability to duplicate modules in a round.

### PR DESCRIPTION
**Pull request by Watermelon914**
## About The Pull Request

Adds the ability to save/load circuits for admins.
Adds the ability to duplicate modules in a round.

## Why It's Good For The Game

Modules are more useful rather than just as an organisation tool because you are able to easily duplicate the code so that you don't have to rewrite the same thing over and over again.

This adds a new machine called the Module Duplicator to all of the circuit labs.

## Changelog
:cl:
expansion: Adds the module duplicator that allows people to duplicate their module components so that they don't have to recreate the entire code again.
admin: Added the ability to save/load integrated circuits.
/:cl:

### CODE REVIEW

**Comment by TemporalOroboros** [comment]
Welp. Circuits were fun while they lasted.

**Comment by Watermelon914** [comment]
> Welp. Circuits were fun while they lasted.

when people don't read pr descriptions or titles

**Comment by coiax** [comment]
DNM :This should only be merged after #60109

File: code/modules/wiremod/duplicator.dm [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+	if(general_data["display_name"])
+		set_display_name(general_data["display_name"])
+
+	var/list/circuit_data = general_data["components"]
+	var/list/identifiers_to_circuit = list()
+	for(var/identifier in circuit_data)
+		var/list/component_data = circuit_data[identifier]
+		var/type = text2path(component_data["type"])
+		if(!ispath(type, /obj/item/circuit_component))
+			LOG_ERROR(errors, "Invalid path for circuit component, expected [/obj/item/circuit_component], got [type]")
```
Comment (Mothblocks) [inline code/modules/wiremod/duplicator.dm:27]: What's with the formatting here, on `[/obj/item/circuit_component]`?

File: code/modules/wiremod/duplicator.dm [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+/obj/item/integrated_circuit/proc/convert_to_json()
+	var/list/circuit_to_identifiers = list()
+	var/list/identifiers = list()
+	var/list/external_objects = list() // Objects that are connected to a component. These objects will be linked to the components.
+	for(var/obj/item/circuit_component/component as anything in attached_components)
+		var/index = 1
+		var/identifier = "[component.type][index]"
+		while(identifier in identifiers)
+			index++
+			identifier = "[component.type][index]"
```
Comment (Mothblocks) [inline code/modules/wiremod/duplicator.dm:102]:
```suggestion
		var/identifier = "[component.type][identifiers.len + 1]"
```

**Review by Mothblocks — CHANGES_REQUESTED** [review]
Cool stuff, a few nitpicks.

File: code/modules/wiremod/duplicator.dm (hunk 1 above)
Comment (Watermelon914) [reply code/modules/wiremod/duplicator.dm:27]: it was basically a way to make sure that if this type somehow doesn't get replaced if the type gets removed, it throws an error. I can change it to just be a string but I thought it'd be more sane to do it like this.

Comment (Mothblocks) [reply code/modules/wiremod/duplicator.dm:27]: Makes sense.

**Pull request merged**